### DISCUSSION TITLE
cryspy.stat のNone

### DISCUSSION THREAD

**Discussion by nim-hrkn** [Tomoki-YAMASHITA/CrySPY#31] [General]
書き換えていてエラーが起きたので、
cryspy.statにNoneやFalseを加えるのは意図した動作ですか。

<pre>
[structure]
struc_mode = crystal
natot = 8
atype = Ga N
nat = 4 4
mol_file = None <---
...
[option]
stop_chkpt = 0
load_struc_flag = False
stop_next_struc = False
recalc = 
append_struc_ea = False
energy_step_flag = False <---
struc_step_flag = False
force_step_flag = False
stress_step_flag = False
</pre>

**Comment by Tomoki-YAMASHITA** [comment]
はい，一応現在使用しているインプット変数をcryspy.statに書き出すということをしています．
裏で（デフォルトで）設定されているものも含めて書き出しています．

続きからcryspyを始めるとき，cryspy.inを前回から変更したかどうかチェックするために
cryspy.statを読み込んで前回使ったインプット変数を取得，比較しています．

特にstep_data関係（最適化の軌跡データ）は途中でTrueなどにするとエラーが出るようになっています．
（途中で変更するとデータの整合性が取れなくなる可能性があるため）

**Reply by nim-hrkn** [reply]
cryspy.inとcryspy.statとの整合性を見ている、でいいでしょうか。
細かいことですが、readin.pyで
<pre>
    try:
        energy_step_flag = config.getboolean('option', 'energy_step_flag')
        if calc_code in ['LAMMPS', 'OMX']:
            raise NotImplementedError('energy_step_flag: only VASP, QE, and soiap for now')
    except (configparser.NoOptionError, configparser.NoSectionError):
        energy_step_flag = False
</pre>
としていますが、最後で全てに対して
        fout.write('append_struc_ea = {}\n'.format(append_struc_ea))
        fout.write('energy_step_flag = {}\n'.format(energy_step_flag))
        fout.write('struc_step_flag = {}\n'.format(struc_step_flag))
        fout.write('force_step_flag = {}\n'.format(force_step_flag))
        fout.write('stress_step_flag = {}\n'.format(stress_step_flag))
        fout.write('\n\n')
としています。

Noneも同じで、余分な=Noneのキーワードが入っています。
ConfigParserでNoneは文字列として読み込まれるのでは無いかと思います。

**Reply by Tomoki-YAMASHITA** [reply]
はい，前回のインプット(cryspy.stat）と今回のインプット（cryspy.in）の整合性をチェックして，変えてはいけないところを変えた場合にエラーが出せるようにしています．

``` python
fout.write('append_struc_ea = {}\n'.format(append_struc_ea))
fout.write('energy_step_flag = {}\n'.format(energy_step_flag))
fout.write('struc_step_flag = {}\n'.format(struc_step_flag))
fout.write('force_step_flag = {}\n'.format(force_step_flag))
fout.write('stress_step_flag = {}\n'.format(stress_step_flag))
fout.write('\n\n')
```
ここは cryspy.outに初回のインプットを書き出すところですね．あまり意味はないです．

cryspy.statの情報は

``` python
def diffinstat(stat):
    logic_change = False

    # ---------- old input
    # ------ basic
    old_algo = stat.get('basic', 'algo')
    old_calc_code = stat.get('basic', 'calc_code')
    old_tot_struc = stat.getint('basic', 'tot_struc')
    if not calc_code == 'ext':
        old_nstage = stat.getint('basic', 'nstage')
        old_njob = stat.getint('basic', 'njob')
        old_jobcmd = stat.get('basic', 'jobcmd')
        old_jobfile = stat.get('basic', 'jobfile')

省略
```

このあたりで読んでいて
configparserでNoneは文字列で読んで，そのあと`None`に変換しています．

```
    if old_mol_file == 'None':
        old_mol_file = None    # character --> None
    else:
        old_mol_file = [a for a in old_mol_file.split()]    # list
    old_nmol = stat.get('structure', 'nmol')
    if old_nmol == 'None':
        old_nmol = None    # character --> None
    else:
        old_nmol = [int(x) for x in old_nmol.split()]    # str --> int list

```


True/Falseのboolで表現できない上記のようなところはNoneを使っていたりします．
